man/ion.1.in
.TH ION 1
.SH NAME
Ion - an X11 window manager
.SH SYNOPSIS
.B ion
.I "[options]"
.SH "DESCRIPTION"

Ion is a tiling tabbed window manager designed with keyboard users in mind.
 
.SH "OPTIONS"
.TP
.B \-display host:display.screen
X display to manage
.TP
.B \-cfgfile configfile
Config file to use instead of default
.TP
.B \-moduledir dir
Directory to search for module
.TP
.B \-confdir dir
Directory to search for configuration files and other scripts
.TP
.B \-oneroot
On X servers with multiple (traditional non-Xinerama) screens, manage only
default screen (root window), not all, as given by the -display option or
in the DISPLAY environment variable.
.TP
.B \-sessionname session_name
Set sesion name. This option affects where workspace and other savefiles are
put (~/.ion2/session_name if option set).
.TP
.B \-xinerama 0|1
If Ion was compiled with Xinerama support, this option can be used to
enable/disable (1/0) the use of it. The default is to use Xinerama
screen information.
.TP
.B \-i18n
Enable use of multibyte string routines, actual encoding depending on the
locale.
.TP
.B \-help
Show help on command line options
.TP
.B \-version
Show version
.TP
.B \-about
Show about text (version, copyright)

.SH BASIC CONCEPTS

This section is an overview of the types objects that appear on an X
display managed by Ion and their relationships in a standard setup.
This information is necessary to understand the operations of different
key bindings. For a more detailed explanation, needed for writing
custom bindings configuration files, see the additional documentation
available from the Ion Web page.

The top-level objects that matter are screens and they correspond to
physical screens. Screens contain workspaces and client windows put in
full screen mode. Only one of these "multiplexed" objects can be viewed
at a time. 

Two types of workspaces currently supported; tiled (traditional Ion-style) 
and floating (pwm-style). Workspaces are not visible objects but rather 
collections of frames. Frames are the objects with decorations such as 
tabs and borders and there are two types of these corresponding to the
diffent types of workspaces.

Frames contain (multiplex) client windows, to each of which corresponds a
tab in the frame's decoration. Only one client window (or other object) 
can be shown at a time in each frame. (In a traditional window manager
each frame may contain exactly one client window.)

"Queries" are boxes that appear at the bottoms of frames or screens 
and are used to ask the user for textual input. Most queries support 
tab-completion.


.SH KEYS

These are the default keybindings.
.B Mod1
depends on your system. On PC:s with XFree86 it is probably bound to the left
alt key (\fBAlt_L\fP).
On Suns it is usually bound to the diamond keys (\fBMeta_L\fP, \fBMeta_R\fP). Use
\fIxmodmap(1x)\fP to find out.

.SS F-keys

.TP
.B F1
Show manual page (query, uses ion-man). Tab-completion can be enabled by
specifying a list of directories in the variable query_man_path.
.TP
.B Mod1+F1
Show Ion manual page.
.TP
.B F2
Run an xterm.
.TP
.B F3
Run a program (query). Precede the program name with a colon (:) to run in xterm (ion-runinxterm).
.TP
.B Mod1+F3
Open a Lua code query. You can call all the functions listed in the
Ion configuration document (see section SEE ALSO) here. The 
variable "_" is bound to the frame in which the query was opened.
.TP
.B F4
Start new SSH session (query, uses ion-ssh). Tab-completion of hosts can be
enabled by specifying them in the configuration variable query_ssh_hosts.
.TP
.B F5
Open a file for editing (query). Uses ion-edit which by default uses run-mailcap.
.TP
.B F6
Open a file for viewing (query). Uses ion-view which by default uses run-mailcap.
.TP
.B F9
Go to existing workspace or create new (query).
.TP
.B Mod1+F9
Create a new tiled workspace with a name of the form WIonWS<n>.
.TP
.B F12
Main menu

.SS Navigation

.TP
.B Mod1+n
Go to n:th workspace (n=1..9, 0)
.TP
.B Mod1+Right
Go to next workspace
.TP
.B Mod1+Left
Go to previous workspace
.TP
.B Mod1+Shift+n
Go to n:th screen (n=1,2)
.TP
.B Mod1+Shift+Right
Go to next screen
.TP
.B Mod1+Shift+Left
Go to previous screen
.TP
.B Mod1+G
Go to client (query; only works in a frame)
.TP
.B Mod1+K K
Go to previously active object



.IP "\fINavigation on tiled workspaces\fR"

.TP
.B Mod1+N
Go to below frame
.TP
.B Mod1+P
Go to above frame
.TP
.B Mod1+Tab
Go to right frame
.TP
.B Mod1+K Tab
Go to left frame

.IP "\fINavigation on floating workspaces\fR"

.TP
.B Mod1+Tab
Circulate
.TP
.B Mod1+K Tab
Backcirculate

.SS Closing and destroying objects.

.TP
.B Mod1+C
Close active object if possible
.TP
.B Mod1+K C
Kill client window
.TP
.B Mod1+K X
Destroy frame

.SS Frame operations

.TP
.B Mod1+K N
Go to next client window within frame
.TP
.B Mod1+K P
Go to previous client window within frame
.TP
.B Mod1+K n
Go to n:th client window within frame (n=1..9, 0)
.TP
.B Mod1+A
Attach named client window in the frame (query)
.TP
.B Mod1+M
Frame context menu


.IP "\fITags\fR"

.TP
.B Mod1+T
Toggle current client window's tag state (indicated by a right-angle
in the top-right corner of the window's tab)
.TP
.B Mod1+K T
Clear all objects' tag state
.TP
.B Mod1+K A
Attach all tagged windows to the frame

.IP "\fIResizing and splitting\fR"

.TP
.B Mod1+K V
Maximize vertically
.TP
.B Mod1+K H
Maximize horizontally
.TP
.B Mod1+S
Split vertically (tiled workspaces only)
.TP
.B Mod1+K S
Split horizontally (tiled workspaces only)
.TP
.B Mod1+R
Enter resize mode

.IP "\fIResize mode bindings\fR"

.TP
.B Left, Right, Up Down or F, B, P, N
Grow the frame in the specific direction
.TP
.B Shift+Left, Right, Up Down or Shift+F, B, P, N
Shrink the frame in the specific direction
.TP
.B Mod1+Left, Right, Up Down or Mod1+F, B, P, N
On floating frame worskacpes: move the frame in the specific direction.

.SS Miscellaneous
.TP
.B Mod1+L
Broken application resize kludge; try to make that broken application
that didn't get it the first time to understand the real size of the
active window
.TP
.B Mod1+K Q
Send next key press to active client window.


.SH LINE-EDITING KEYS

These keys can be used in the line editor (queries). They are similar to
those of \fIjoe(1)\fP.
Selection works differently, though, while the keys are equivalent.

.SS Movement

.TP
.B Control+F, Right arrow
Go forward character
.TP
.B Control+B, Left arrow
Go backward character
.TP
.B Control+E, End
Go to end of line
.TP
.B Control+A, Home
Go to beginning of line
.TP
.B Control+Z
Backward skip word
.TP
.B Control+X
Forward skip word

.SS Delete

.TP
.B Control+D, Delete
Delete next character
.TP
.B Control+H, Backspace
Delete previous character
.TP
.B Control+J
Delete to end of line
.TP
.B Control+Y
Delete whole line
.TP
.B Control+W
Delete word
.TP
.B Control+O
Backward delete word

.SS Selections

.TP
.B Control+K B
Set mark (begin selection)
.TP
.B Control+K K
Copy selection
.TP
.B Control+K Y
Cut selection
.TP
.B Control+K C
Paste

.SS Completion and history

.TP
.B Tab
Complete or list possible completions (if available)
.TP
.B Control+U, Page up
Scroll completions up
.TP
.B Control+V, Page down
Scroll completions down
.TP
.B Control+P, Up arrow
Previous history entry
.TP
.B Control+N, Down arrow
Next history entry

.SS Cancel, accept

.TP
.B Control+C, Escape
Cancel
.TP
.B Control+M, Enter
Accept

.SH POINTING DEVICE BUTTONS

The bindings for the two types of frames are again different because not
all legacy/floating frame actions are available on tiled frames and thus
those buttons have been mapped to other actions.

.SS Tiled frames

.TP
.B Button1 (left) or Button2 (middle) on tab
Click: switch client, drag: drag&drop clients between frames
.TP
.B Button1 (left) on border
Drag: resize
.TP
.B Mod1+Button1 (left) or Mod1+Button3 (right)
Resize.
.TP
.B Button3 (right) on tab
Context menu

.SS Floating frames

.TP
.B Button1 (left)
Press: raise frame, click tab: switch client, double-click tab: toggle
shade mode, drag border: resize, drag elsewhere: move frame
.TP
.B Button2 (middle) on tab
Click: switch client, drag: drag&drop clients between frames
.TP
.B Mod1+Button1 (left)
Click: raise, drag: move
.TP
.B Mod1+Button3 (right)
Click: lower, drag: resize
.TP
.B Button3 (right) on tab
Context menu

.SH FILES AND DIRECTORIES
.TP
.B ETCDIR/
System default configuration files
.TP
.B ETCDIR/ion.lua
System default main configuratio files
.TP
.B ETCDIR/look-*.lua
Colour scheme configuration files
.TP
.B ~/.ion2/
User configuration files
.TP
.B ~/.ion2/ion.lua
User default configuration file (overrides system default)

.SH SEE ALSO
The Ion home page, \fIhttp://iki.fi/tuomov/ion/\fP
.PP
The document "Ion: Configuring and extending with Lua" found on the
Ion home page.
.PP
.I DOCDIR/
.PP
\fIX(1x)\fP, \fIpwm(1)\fP

.SH AUTHOR
Ion was written by Tuomo Valkonen <tuomov at iki.fi>.


CHANGELOG
Version 20140528:
* Improvement: 'windows' key is now the default META key

* Bugfix: seldomly triggered crash in memory pool
* Bugfix: save the session atomically
* Bugfix: various fixes to contributed scripts, notably more lua5.2 support

* Improvement: various code cleanups (thanks to Campbell Barton)
* Improvement: various build/packaging improvements (thanks to Dima Kogan)

* Feature: Workspace indicator (thanks to Dima Kogan)
* Feature: Delayed population of the focuslist (thanks to Dima Kogan)

* Improvement: improved kludges/focus behavior (IntelliJ IDEA, xeyes)
* Improvement: expose 'warp_pointer' to lua
* Improvement: lock using i3lock when available


ion/ion.c
/*
 * ion/ion/ion.c
 *
 * Copyright (c) Tuomo Valkonen 1999-2004. 
 *
 * Ion is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <fcntl.h>

#include <libtu/util.h>
#include <libtu/optparser.h>
#include <libtu/errorlog.h>

#include <ioncore/common.h>
#include <ioncore/global.h>
#include <ioncore/ioncore.h>
#include <ioncore/readconfig.h>
#include <ioncore/exec.h>
#include <ioncore/mainloop.h>
#include "../version.h"


#ifndef CF_NEW_USER_MESSAGE
#define CF_NEW_USER_MESSAGE SHAREDIR"/welcome_message.txt"
#endif


/* Options. Getopt is not used because getopt_long is quite gnu-specific
 * and they don't know of '-display foo' -style args anyway.
 * Instead, I've reinvented the wheel in libtu :(.
 */
static OptParserOpt ion_opts[]={
    {OPT_ID('d'),     "display",     OPT_ARG, "host:dpy.scr", "X display to use"},
    {'c',             "conffile", OPT_ARG, "config_file", "Configuration file"},
    {OPT_ID('o'),     "oneroot",  0, NULL, "Manage default root window/non-Xinerama screen only"},
    {OPT_ID('c'),     "confdir",     OPT_ARG, "dir", "Search directory for configuration files"},
    {OPT_ID('l'),     "moduledir", OPT_ARG, "dir", "Search directory for modules"},
#ifndef CF_NOXINERAMA    
    {OPT_ID('x'),     "xinerama", OPT_ARG, "1|0", "Use Xinerama screen information (default: 1/yes)"},
#else
    {OPT_ID('x'),     "xinerama", OPT_ARG, "?", "Ignored: not compiled with Xinerama support"},
#endif
    {OPT_ID('s'),   "sessionname", OPT_ARG, "session_name", "Name of session (affects savefiles)"},
    {OPT_ID('i'),   "i18n", 0, NULL, "Enable use of multibyte string routines, actual "
                                     "encoding depending on the locale."},
    END_OPTPARSEROPTS
};


static const char ion_usage_tmpl[]=
    "Usage: $p [options]\n\n$o\n";


static OptParserCommonInfo ion_cinfo={
    ION_VERSION,
    ion_usage_tmpl,
    NULL,
};



void check_new_user_help()
{
    const char *userdir=ioncore_userdir();
    char *oldbeard=NULL;
    pid_t pid;

    if(userdir==NULL){
        warn("Could not get user configuration file directory.");
        return;
    }
    
    libtu_asprintf(&oldbeard, "%s/.welcome_msg_displayed", userdir);
    
    if(oldbeard==NULL){
        warn_err();
        return;
    }
    
    if(access(oldbeard, F_OK)==0){
        free(oldbeard);
        return;
    }

    if(!ioncore_exec(CF_XMESSAGE" "CF_NEW_USER_MESSAGE))
        return;

    /* This should actually be done when less or xmessage returns,
     * but that would mean yet another script...
     */
    mkdir(userdir, 0700);
    if(open(oldbeard, O_CREAT|O_RDWR, 0600)<0)
        warn_err_obj(oldbeard);
}


int main(int argc, char*argv[])
{
    const char *cfgfile="ion";
    const char *display=NULL;
    char *cmd=NULL;
    int stflags=0;
    int opt;
    ErrorLog el;
    FILE *ef=NULL;
    char *efnam=NULL;
    bool may_continue=FALSE;
    bool i18n=FALSE;
    
    libtu_init(argv[0]);

    if(!ioncore_init(argc, argv))
        return EXIT_FAILURE;

    ion_cinfo.about=ioncore_aboutmsg();
    
    ioncore_add_scriptdir(EXTRABINDIR); /* ion-completefile */
    ioncore_add_scriptdir(ETCDIR);
    ioncore_add_scriptdir(SHAREDIR);
    ioncore_add_moduledir(MODULEDIR);
    ioncore_set_userdirs("ion3");

    optparser_init(argc, argv, OPTP_MIDLONG, ion_opts, &ion_cinfo);
    
    while((opt=optparser_get_opt())){
        switch(opt){
        case OPT_ID('d'):
            display=optparser_get_arg();
            break;
        case 'c':
            cfgfile=optparser_get_arg();
            break;
        case OPT_ID('c'):
            ioncore_add_scriptdir(optparser_get_arg());
            break;
        case OPT_ID('l'):
            ioncore_add_moduledir(optparser_get_arg());
            break;
        case OPT_ID('o'):
            stflags|=IONCORE_STARTUP_ONEROOT;
            break;
        case OPT_ID('x'):
            {
                const char *p=optparser_get_arg();
                if(strcmp(p, "1")==0)
                    stflags&=~IONCORE_STARTUP_NOXINERAMA;
                else if(strcmp(p, "0")==0)
                    stflags|=IONCORE_STARTUP_NOXINERAMA;
                else
                    warn("Invalid parameter to -xinerama.");
            }
            break;
        case OPT_ID('s'):
            ioncore_set_sessiondir(optparser_get_arg());
            break;
        case OPT_ID('i'):
            i18n=TRUE;
            break;
        default:
            optparser_print_error();
            return EXIT_FAILURE;
        }
    }

    /* We may have to pass the file to xmessage so just using tmpfile()
     * isn't sufficient.
     */
    libtu_asprintf(&efnam, "%s/ion-%d-startup-errorlog", P_tmpdir,
                   getpid());
    if(efnam==NULL){
        warn_err("Failed to create error log file");
    }else{
        ef=fopen(efnam, "wt");
        if(ef==NULL){
            warn_err_obj(efnam);
            free(efnam);
            efnam=NULL;
        }
        fprintf(ef, "Ion startup error log:\n");
        errorlog_begin_file(&el, ef);
    }

    /* Set up locale and detect encoding.
     */
    if(i18n)
        ioncore_init_i18n();
    
    if(ioncore_startup(display, cfgfile, stflags))
        may_continue=TRUE;

fail:
    if(!may_continue)
        warn("Refusing to start due to encountered errors.");
    else
        check_new_user_help();
    
    if(ef!=NULL){
        pid_t pid=-1;
        if(errorlog_end(&el)){
            fclose(ef);
            pid=fork();
            if(pid==0){
                ioncore_setup_environ(DefaultScreen(ioncore_g.dpy));
                if(!may_continue)
                    XCloseDisplay(ioncore_g.dpy);
                else
                    close(ioncore_g.conn);
                libtu_asprintf(&cmd, CF_XMESSAGE " %s", efnam);
                if(system(cmd)==-1)
                    warn_err_obj(cmd);
                unlink(efnam);
                exit(EXIT_SUCCESS);
            }
            if(!may_continue && pid>0)
                waitpid(pid, NULL, 0);
        }else{
            fclose(ef);
        }
        if(pid<0)
            unlink(efnam);
        free(efnam);
    }

    if(!may_continue)
        return EXIT_FAILURE;
    
    ioncore_mainloop();
    
    /* The code should never return here */
    return EXIT_SUCCESS;
}
